flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/batch/connectors/hive/HiveTableSink.java

```diff
@@ -0,0 +1,155 @@
+import org.apache.flink.table.catalog.exceptions.CatalogException;
+import org.apache.flink.table.catalog.hive.client.HiveMetastoreClientFactory;
+import org.apache.flink.table.catalog.hive.client.HiveMetastoreClientWrapper;
+import org.apache.flink.table.catalog.hive.client.HiveShimLoader;
+import org.apache.flink.table.catalog.hive.descriptors.HiveCatalogValidator;
+import org.apache.flink.table.sinks.OutputFormatTableSink;
+import org.apache.flink.table.sinks.TableSink;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.FlinkRuntimeException;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.MetaStoreUtils;
+import org.apache.hadoop.hive.metastore.Warehouse;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.hadoop.hive.metastore.api.StorageDescriptor;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.thrift.TException;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+
+/**
+ * Table sink to write to Hive tables.
+ */
+public class HiveTableSink extends OutputFormatTableSink<Row> {
+
+	private final JobConf jobConf;
+	private final RowTypeInfo rowTypeInfo;
+	private final String dbName;
+	private final String tableName;
+	private final List<String> partitionColumns;
+	private final String hiveVersion;
+
+	// TODO: need OverwritableTableSink to configure this
+	private boolean overwrite = false;
+
+	public HiveTableSink(JobConf jobConf, RowTypeInfo rowTypeInfo, String dbName, String tableName,
+			List<String> partitionColumns) {
+		this.jobConf = jobConf;
+		this.rowTypeInfo = rowTypeInfo;
+		this.dbName = dbName;
+		this.tableName = tableName;
+		this.partitionColumns = partitionColumns;
+		hiveVersion = jobConf.get(HiveCatalogValidator.CATALOG_HIVE_VERSION, HiveShimLoader.getHiveVersion());
+	}
+
+	@Override
+	public OutputFormat<Row> getOutputFormat() {
+		boolean isPartitioned = partitionColumns != null && !partitionColumns.isEmpty();
+		// TODO: need PartitionableTableSink to decide whether it's dynamic partitioning
+		boolean isDynamicPartition = isPartitioned;
+		try (HiveMetastoreClientWrapper client = HiveMetastoreClientFactory.create(new HiveConf(jobConf, HiveConf.class), hiveVersion)) {
+			Table table = client.getTable(dbName, tableName);
+			StorageDescriptor sd = table.getSd();
+			// here we use the sdLocation to store the output path of the job, which is always a staging dir
+			String sdLocation = sd.getLocation();
+			HiveTablePartition hiveTablePartition;
+			if (isPartitioned) {
+				// TODO: validate partition spec
+				// TODO: strip quotes in partition values
+				LinkedHashMap<String, String> strippedPartSpec = new LinkedHashMap<>();
```

Comment:
Is there a reason to make strippedPartSpec `LinkedHashMap` not `Map`?

Reply:
Partition spec requires the partition columns in a specific order. So it's better to use `LinkedHashMap ` to explicitly indicate we need an ordered map here.

Reply:
Makes sense.